Ledc fix

xvil opened the pull request:
# What does this implement/fix?

This PR fixes the bit-depth calculation of LEDC when using the arduino framework.
It also add some fallback mechanism when the initially computed bit-depth is not compatible with the user-defined frequency of the PWM (both for IDF and Arduino)

## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes [link to issue](https://github.com/esphome/issues/issues/4050)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
esphome:
  name: esp-ledc-tst

esp32:
  board: esp32-c3-devkitm-1
  framework:
    type: arduino
      
# Enable logging
logger:
  baud_rate: 115200
  level: verbose
  logs:
    ledc: verbose

# PWM output
output:
  - platform: ledc
    id: fan_pwm
    pin: 
      number: GPIO04
    frequency: 25000 Hz
    inverted: true
    min_power: 0%
    max_power: 100%

# Used to manually set the fan speed
fan:
  - platform: speed
    output: fan_pwm
    name: "Fan Speed"
    id: fan_speed
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

xvil committed:
Fixed bit depth issue + added fallback

xvil committed:
Merge branch 'esphome:dev' into ledc-fix

xvil committed:
Fixed issues reported by Lint Custom

xvil committed:
Fixed issues reported by clang_format script

xvil committed:
Fixed more issues reported by clang_format script

xvil committed:
Fixed more issues reported by clang_format script

xvil committed:
Fixed more issues reported by clang_format script

jesserockz reviewed (changes requested):
I know the existing code doesn't clear the warnings, but if we are making changes, might as well clear the error if it successfully sets the frequency

xvil committed:
Merge branch 'dev' into ledc-fix

xvil committed:
Cleared error status and implemented missing bit_depth fallback in set_frequency for IDF

xvil committed:
Fixed code formatting

xvil committed:
Fixed code formatting

xvil committed:
Fixed formatting

xvil commented:
I have implemented the error status reset.
I have added the fallback of the bit_depth calculation for set_frequency when using the ESP_IDF framework.
The common code has refactored into a dedicated function to avoid duplication as much as possible.

xvil committed:
Merge branch 'esphome:dev' into ledc-fix
